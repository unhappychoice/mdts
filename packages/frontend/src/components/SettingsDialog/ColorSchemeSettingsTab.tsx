import { Box, MenuItem, Select, ToggleButton, ToggleButtonGroup, Typography } from '@mui/material';
import { useTheme } from '@mui/material/styles';
import React, { useCallback } from 'react';
import { DARK_SYNTAX_HIGHLIGHTER_THEMES, LIGHT_SYNTAX_HIGHLIGHTER_THEMES, } from '../../constants';

interface ColorSchemeSettingsTabProps {
  darkMode: 'dark' | 'light' | 'auto';
  syntaxHighlighterTheme: string;
  handleToggleDarkMode: (mode: 'dark' | 'light' | 'auto') => void;
  setSyntaxHighlighterTheme: (theme: string) => void;
}

const ColorSchemeSettingsTab: React.FC<ColorSchemeSettingsTabProps> = ({
  darkMode,
  syntaxHighlighterTheme,
  handleToggleDarkMode,
  setSyntaxHighlighterTheme,
}) => {
  const theme = useTheme();
  const selectableThemes = (darkMode === 'dark' || (darkMode === 'auto' && theme.palette.mode === 'dark'))
    ? DARK_SYNTAX_HIGHLIGHTER_THEMES
    : LIGHT_SYNTAX_HIGHLIGHTER_THEMES;

  const handleDarkModeChange = useCallback((_, newMode) => {
    if (newMode)
      handleToggleDarkMode(newMode);
  }, [handleToggleDarkMode]);

  const handleThemChange = useCallback((e) => {
    setSyntaxHighlighterTheme(e.target.value as string);
  }, [setSyntaxHighlighterTheme]);

  return (
    <Box sx={{ p: 2 }}>
      <Typography variant="subtitle2" sx={{ mb: 2 }}>Dark Mode</Typography>
      <ToggleButtonGroup
        value={darkMode}
        exclusive
        onChange={handleDarkModeChange}
        aria-label="dark mode"
        fullWidth
        size="small"
        sx={{ mb: 2 }}
      >
        <ToggleButton
          value="dark"
          aria-label="dark"
          sx={{
            flexGrow: 1,
            '&.Mui-selected': {
              backgroundColor: theme.palette.primary.main,
              color: theme.palette.primary.contrastText,
              '&:hover': {
                backgroundColor: theme.palette.primary.dark,
              },
            },
          }}
        >
          Dark
        </ToggleButton>
        <ToggleButton
          value="light"
          aria-label="light"
          sx={{
            flexGrow: 1,
            '&.Mui-selected': {
              backgroundColor: theme.palette.primary.main,
              color: theme.palette.primary.contrastText,
              '&:hover': {
                backgroundColor: theme.palette.primary.dark,
              },
            },
          }}
        >
          Light
        </ToggleButton>
        <ToggleButton
          value="auto"
          aria-label="auto"
          sx={{
            flexGrow: 1,
            '&.Mui-selected': {
              backgroundColor: theme.palette.primary.main,
              color: theme.palette.primary.contrastText,
              '&:hover': {
                backgroundColor: theme.palette.primary.dark,
              },
            },
          }}
        >
          Auto (System)
        </ToggleButton>
      </ToggleButtonGroup>

      <Typography variant="subtitle2" sx={{ my: 2 }} id="syntax-highlighter-theme-label">Syntax Highlighter Theme</Typography>
      <Select
        labelId="syntax-highlighter-theme-label"
        id="syntaxHighlighterTheme"
        value={syntaxHighlighterTheme}
        onChange={handleThemChange}
        fullWidth
        variant="outlined"
        size="small"
        sx={{ mb: 1 }}
        data-testid="syntax-highlighter-theme-select"
      >
        <MenuItem value="auto">Auto</MenuItem>
        {selectableThemes.map(theme => (
          <MenuItem key={theme.value} value={theme.value}>
            {theme.name}
          </MenuItem>
        ))}
      </Select>
    </Box>
  );
};

export default ColorSchemeSettingsTab;
