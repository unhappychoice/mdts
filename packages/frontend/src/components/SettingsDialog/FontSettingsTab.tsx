import {
  Box,
  FormControlLabel,
  MenuItem,
  Radio,
  RadioGroup,
  Select,
  Slider,
  TextField,
  Typography
} from '@mui/material';
import React, { useCallback } from 'react';
import { WEB_SAFE_FONTS, WEB_SAFE_MONOSPACE_FONTS, } from '../../constants';

interface FontSettingsTabProps {
  fontSize: number;
  fontFamily: string;
  fontFamilyMonospace: string;
  fontInputMode: 'select' | 'text';
  fontInputModeMonospace: 'select' | 'text';
  setFontSize: (size: number) => void;
  setFontFamily: (font: string) => void;
  setFontFamilyMonospace: (font: string) => void;
  handleFontInputModeChange: (_: React.MouseEvent<HTMLElement>, newMode: 'select' | 'text' | null) => void;
  handleFontInputModeMonospaceChange: (_: React.MouseEvent<HTMLElement>, newMode: 'select' | 'text' | null) => void;
}

const FontSettingsTab: React.FC<FontSettingsTabProps> = ({
  fontSize,
  fontFamily,
  fontFamilyMonospace,
  fontInputMode,
  fontInputModeMonospace,
  setFontSize,
  setFontFamily,
  setFontFamilyMonospace,
  handleFontInputModeChange,
  handleFontInputModeMonospaceChange,
}) => {

  const handleFontSelect = useCallback((e) => {
    setFontFamily(e.target.value as string);
  }, [setFontFamily]);

  const handleFontChange = useCallback((e) => {
    setFontFamily(e.target.value);
  }, [setFontFamily]);

  const handleMonospaceFontSelect = useCallback((e) => {
    setFontFamilyMonospace(e.target.value as string);
  }, [setFontFamilyMonospace]);

  const handleMonospaceFontChange = useCallback((e) => {
    setFontFamilyMonospace(e.target.value);
  }, [setFontFamilyMonospace]);

  const handleChangeFontSize = useCallback((_, newValue) => {
    setFontSize(newValue as number);
  }, [setFontSize]);

  return (
    <Box sx={{ p: 2 }}>
      <Typography variant="subtitle2" gutterBottom>Font Size</Typography>
      <Slider
        value={fontSize}
        onChange={handleChangeFontSize}
        aria-label="Font Size"
        aria-labelledby="font-size-slider"
        valueLabelDisplay="auto"
        step={1}
        marks
        min={10}
        max={24}
        size="small"
        sx={{ mb: 1 }}
      />

      <Typography variant="subtitle2" gutterBottom sx={{ mt: 2 }} id="font-family-label">Font Family</Typography>
      <RadioGroup
        row
        aria-labelledby="font-family-label"
        name="font-input-mode-group"
        value={fontInputMode}
        onChange={handleFontInputModeChange}
        sx={{ mb: 1 }}
      >
        <FormControlLabel value="select" control={<Radio size="small" />} label="Select from list" />
        <FormControlLabel value="text" control={<Radio size="small" />} label="Enter manually" />
      </RadioGroup>

      {fontInputMode === 'select' ? (
        <Select
          labelId="font-family-label"
          id="fontFamily"
          value={fontFamily}
          onChange={handleFontSelect}
          fullWidth
          variant="outlined"
          size="small"
          sx={{ mb: 1 }}
          data-testid="font-family-select"
        >
          {WEB_SAFE_FONTS.map((font) => (
            <MenuItem key={font} value={font} style={{ fontFamily: font }}>
              {font}
            </MenuItem>
          ))}
        </Select>
      ) : (
        <TextField
          autoFocus
          margin="dense"
          id="fontFamily"
          aria-labelledby="font-family-label"
          type="text"
          fullWidth
          variant="outlined"
          size="small"
          value={fontFamily}
          sx={{ mt: 0, mb: 1 }}
          onChange={handleFontChange}
          data-testid="font-family-input"
        />
      )}

      <Typography variant="subtitle2" gutterBottom sx={{ mt: 2 }} id="monospace-font-family-label">Monospace Font Family</Typography>
      <RadioGroup
        row
        aria-labelledby="monospace-font-family-label"
        name="monospace-font-input-mode-group"
        value={fontInputModeMonospace}
        onChange={handleFontInputModeMonospaceChange}
        sx={{ mb: 1 }}
      >
        <FormControlLabel value="select" control={<Radio size="small" />} label="Select from list" />
        <FormControlLabel value="text" control={<Radio size="small" />} label="Enter manually" />
      </RadioGroup>

      {fontInputModeMonospace === 'select' ? (
        <Select
          labelId="monospace-font-family-label"
          id="fontFamilyMonospace"
          value={fontFamilyMonospace}
          onChange={handleMonospaceFontSelect}
          fullWidth
          variant="outlined"
          size="small"
          sx={{ mb: 1 }}
          data-testid="monospace-font-family-select"
        >
          {WEB_SAFE_MONOSPACE_FONTS.map((font) => (
            <MenuItem key={font} value={font} style={{ fontFamily: font }}>
              {font}
            </MenuItem>
          ))}
        </Select>
      ) : (
        <TextField
          margin="dense"
          id="fontFamilyMonospace"
          aria-labelledby="monospace-font-family-label"
          type="text"
          fullWidth
          variant="outlined"
          size="small"
          value={fontFamilyMonospace}
          sx={{ mt: 0, mb: 1 }}
          onChange={handleMonospaceFontChange}
          data-testid="monospace-font-family-input"
        />
      )}
    </Box>
  );
};

export default FontSettingsTab;
