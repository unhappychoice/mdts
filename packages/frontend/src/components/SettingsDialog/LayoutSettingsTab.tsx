import {
  Box,
  ToggleButton,
  ToggleButtonGroup,
  Typography
} from '@mui/material';
import { useTheme } from '@mui/material/styles';
import React, { useCallback } from 'react';

interface LayoutSettingsTabProps {
  contentMode: 'full' | 'compact';
  handleToggleContentMode: (mode: 'full' | 'compact') => void;
}

const LayoutSettingsTab: React.FC<LayoutSettingsTabProps> = ({ contentMode, handleToggleContentMode }) => {
  const theme = useTheme();

  const handleChange = useCallback((_, newMode) => {
    if (newMode)
      handleToggleContentMode(newMode);
  } , [handleToggleContentMode]);

  return (
    <Box sx={{ p: 2 }}>
      <Typography variant="subtitle2" sx={{ mb: 2 }}>Content Width</Typography>
      <ToggleButtonGroup
        value={contentMode}
        exclusive
        onChange={handleChange}
        aria-label="content width"
        fullWidth
        size="small"
        sx={{ mb: 2 }}
      >
        <ToggleButton
          value="full"
          aria-label="full"
          sx={{
            flexGrow: 1,
            '&.Mui-selected': {
              backgroundColor: theme.palette.primary.main,
              color: theme.palette.primary.contrastText,
              '&:hover': {
                backgroundColor: theme.palette.primary.dark,
              },
            },
          }}
        >
          Full
        </ToggleButton>
        <ToggleButton
          value="compact"
          aria-label="compact"
          sx={{
            flexGrow: 1,
            '&.Mui-selected': {
              backgroundColor: theme.palette.primary.main,
              color: theme.palette.primary.contrastText,
              '&:hover': {
                backgroundColor: theme.palette.primary.dark,
              },
            },
          }}
        >
          Compact
        </ToggleButton>
      </ToggleButtonGroup>
    </Box>
  );
};

export default LayoutSettingsTab;
