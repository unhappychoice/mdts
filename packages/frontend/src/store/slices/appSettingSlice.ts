import { createSlice } from '@reduxjs/toolkit';

type DarkMode = 'dark' | 'light' | 'auto';
type ContentMode = 'full' | 'compact';

interface AppSettingState {
  darkMode: DarkMode;
  contentMode: ContentMode;
  fileTreeOpen: boolean;
  outlineOpen: boolean;
}

const initialState: AppSettingState = {
  darkMode: 'auto',
  contentMode: 'compact',
  fileTreeOpen: true,
  outlineOpen: true,
};

const appSettingSlice = createSlice({
  name: 'appSetting',
  initialState,
  reducers: {
    saveAppSetting: (state, action: PayloadAction<{ darkMode: DarkMode, contentMode: ContentMode }>) => {
      state.darkMode = action.payload.darkMode;
      state.contentMode = action.payload.contentMode;
    },
    toggleFileTree: (state) => {
      state.fileTreeOpen = !state.fileTreeOpen;
    },
    toggleOutline: (state) => {
      state.outlineOpen = !state.outlineOpen;
    },
  },
});

export const { saveAppSetting, toggleFileTree, toggleOutline } = appSettingSlice.actions;

export default appSettingSlice.reducer;
