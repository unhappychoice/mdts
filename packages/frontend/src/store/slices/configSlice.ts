import { createAsyncThunk, createSlice, PayloadAction } from '@reduxjs/toolkit';
import { fetchData } from '../../api';

export const fetchConfig = createAsyncThunk('config/fetchConfig', async () => {
  return await fetchData<ConfigState>('/api/config', 'json');
});

export const saveConfigToBackend = createAsyncThunk(
  'config/saveConfigToBackend',
  async (config: ConfigState, { dispatch }) => {
    try {
      await fetch('/api/config', {
        method: 'POST',
        headers: {
          'Content-Type': 'application/json',
        },
        body: JSON.stringify(config),
      });
      dispatch(fetchConfig()); // Re-fetch config after saving
    } catch (error) {
      console.error('Failed to save settings:', error);
      throw error; // Re-throw to allow error handling in component
    }
  }
);

export interface ConfigState {
  fontFamily: string;
  fontFamilyMonospace: string;
  fontSize: number;
  syntaxHighlighterTheme: string;
}

const initialState: ConfigState = {
  fontFamily: 'Roboto',
  fontFamilyMonospace: 'monospace',
  fontSize: 14,
  syntaxHighlighterTheme: 'auto',
};

const configSlice = createSlice({
  name: 'config',
  initialState,
  reducers: {},
  extraReducers: (builder) => {
    builder.addCase(fetchConfig.fulfilled, (state, action: PayloadAction<ConfigState>) => {
      state.fontFamily = action.payload.fontFamily;
      state.fontFamilyMonospace = action.payload.fontFamilyMonospace;
      state.fontSize = action.payload.fontSize;
      state.syntaxHighlighterTheme = action.payload.syntaxHighlighterTheme;
    });
  },
});

export default configSlice.reducer;
