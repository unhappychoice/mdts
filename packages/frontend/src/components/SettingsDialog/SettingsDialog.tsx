import { Box, Button, Dialog, DialogActions, DialogContent, DialogTitle, Tab, Tabs, } from '@mui/material';
import { useTheme } from '@mui/material/styles';
import React, { useCallback, useEffect, useState } from 'react';
import { useDispatch, useSelector } from 'react-redux';
import { WEB_SAFE_FONTS, WEB_SAFE_MONOSPACE_FONTS, } from '../../constants';
import { saveAppSetting } from '../../store/slices/appSettingSlice';
import { saveConfigToBackend } from '../../store/slices/configSlice';
import { AppDispatch, RootState } from '../../store/store';
import ColorSchemeSettingsTab from './ColorSchemeSettingsTab';
import FontSettingsTab from './FontSettingsTab';
import LayoutSettingsTab from './LayoutSettingsTab';
import TabPanel from './TabPanel';

interface SettingsDialogProps {
  open: boolean;
  onClose: () => void;
}

const SettingsDialog: React.FC<SettingsDialogProps> = ({ open, onClose }) => {
  const dispatch: AppDispatch = useDispatch();
  const theme = useTheme();
  const {
    darkMode: initialDarkMode,
    contentMode: initialContentMode
  } = useSelector((state: RootState) => state.appSetting);
  const {
    fontFamily: initialFontFamily,
    fontFamilyMonospace: initialFontFamilyMonospace,
    fontSize: initialFontSize,
    syntaxHighlighterTheme: initialSyntaxHighlighterTheme
  } = useSelector((state: RootState) => state.config);

  const [selectedTab, setSelectedTab] = useState(0);
  const [darkMode, setDarkMode] = useState(initialDarkMode);
  const [contentMode, setContentMode] = useState(initialContentMode);
  const [fontFamily, setFontFamily] = useState(initialFontFamily);
  const [fontFamilyMonospace, setFontFamilyMonospace] = useState(initialFontFamilyMonospace);
  const [fontSize, setFontSize] = useState(initialFontSize);
  const [syntaxHighlighterTheme, setSyntaxHighlighterTheme] = useState(initialSyntaxHighlighterTheme);
  const [fontInputMode, setFontInputMode] = useState<'select' | 'text'>(WEB_SAFE_FONTS.includes(initialFontFamily) ? 'select' : 'text');
  const [fontInputModeMonospace, setFontInputModeMonospace] = useState<'select' | 'text'>(WEB_SAFE_MONOSPACE_FONTS.includes(initialFontFamilyMonospace) ? 'select' : 'text');

  useEffect(() => {
    setContentMode(initialContentMode);
    setDarkMode(initialDarkMode);
    setSyntaxHighlighterTheme(initialSyntaxHighlighterTheme);
    setFontFamily(initialFontFamily);
    setFontFamilyMonospace(initialFontFamilyMonospace);
    setFontSize(initialFontSize);
    setFontInputMode(WEB_SAFE_FONTS.includes(initialFontFamily) ? 'select' : 'text');
    setFontInputModeMonospace(WEB_SAFE_MONOSPACE_FONTS.includes(initialFontFamilyMonospace) ? 'select' : 'text');
  }, [
    initialContentMode,
    initialDarkMode,
    initialSyntaxHighlighterTheme,
    initialFontFamily,
    initialFontFamilyMonospace,
    initialFontSize
  ]);

  const handleSave = useCallback(() => {
    dispatch(saveConfigToBackend({ fontFamily, fontFamilyMonospace, fontSize, syntaxHighlighterTheme }));
    dispatch(saveAppSetting({ darkMode, contentMode }));

    onClose();
  }, [contentMode, darkMode, dispatch, fontFamily, fontFamilyMonospace, fontSize, onClose, syntaxHighlighterTheme]);

  const handleCancel = useCallback(() => {
    setDarkMode(initialDarkMode);
    setContentMode(initialContentMode);
    setFontFamily(initialFontFamily);
    setFontFamilyMonospace(initialFontFamilyMonospace);
    setFontSize(initialFontSize);
    setSyntaxHighlighterTheme(initialSyntaxHighlighterTheme);
    setFontInputMode(WEB_SAFE_FONTS.includes(initialFontFamily) ? 'select' : 'text');
    setFontInputModeMonospace(WEB_SAFE_MONOSPACE_FONTS.includes(initialFontFamilyMonospace) ? 'select' : 'text');
    onClose();
  }, [
    initialContentMode,
    initialDarkMode,
    initialFontFamily,
    initialFontFamilyMonospace,
    initialFontSize,
    initialSyntaxHighlighterTheme,
    onClose
  ]);

  const handleSelectTab = useCallback((_, newValue) => {
    setSelectedTab(newValue);
  }, [setSelectedTab]);

  const handleToggleDarkMode = useCallback((mode: 'dark' | 'light' | 'auto') => {
    setDarkMode(mode);
  }, []);

  const handleToggleContentMode = useCallback((mode: 'full' | 'compact') => {
    setContentMode(mode);
  }, []);

  const handleFontInputModeChange = useCallback((_: React.MouseEvent<HTMLElement>, newMode: 'select' | 'text' | null) => {
    if (newMode) {
      setFontInputMode(newMode);
      if (newMode === 'select' && !WEB_SAFE_FONTS.includes(fontFamily)) {
        setFontFamily(WEB_SAFE_FONTS[0]); // Reset to first web-safe font if current is not in list
      }
    }
  }, [fontFamily]);

  const handleFontInputModeMonospaceChange = useCallback((_: React.MouseEvent<HTMLElement>, newMode: 'select' | 'text' | null) => {
    if (newMode) {
      setFontInputModeMonospace(newMode);
      if (newMode === 'select' && !WEB_SAFE_MONOSPACE_FONTS.includes(fontFamilyMonospace)) {
        setFontFamilyMonospace(WEB_SAFE_MONOSPACE_FONTS[0]); // Reset to first web-safe monospace font if current is not in list
      }
    }
  }, [fontFamilyMonospace]);

  const handleReset = useCallback(() => {
    const defaultDarkMode = 'auto';
    const defaultContentMode = 'compact';
    const defaultFontFamily = 'Roboto';
    const defaultFontFamilyMonospace = 'monospace';
    const defaultFontSize = 14;
    const defaultSyntaxHighlighterTheme = 'auto';

    setDarkMode(defaultDarkMode);
    setContentMode(defaultContentMode);
    setFontFamily(defaultFontFamily);
    setFontFamilyMonospace(defaultFontFamilyMonospace);
    setFontSize(defaultFontSize);
    setSyntaxHighlighterTheme(defaultSyntaxHighlighterTheme);
    setFontInputMode(WEB_SAFE_FONTS.includes(defaultFontFamily) ? 'select' : 'text');
    setFontInputModeMonospace(WEB_SAFE_MONOSPACE_FONTS.includes(defaultFontFamilyMonospace) ? 'select' : 'text');

    dispatch(saveConfigToBackend({
      fontFamily: defaultFontFamily,
      fontFamilyMonospace: defaultFontFamilyMonospace,
      fontSize: defaultFontSize,
      syntaxHighlighterTheme: defaultSyntaxHighlighterTheme
    }));
    dispatch(saveAppSetting({ darkMode: defaultDarkMode, contentMode: defaultContentMode }));
  }, [dispatch]);

  return (
    <Dialog maxWidth='lg' open={open} onClose={onClose}>
      <DialogTitle sx={{ pt: 2, pb: 2, borderBottom: '1px solid', borderColor: 'divider' }}>Appearance Settings</DialogTitle>
      <DialogContent sx={{ width: 800, height: 600, p: 0 }} className="custom-scrollbar">
        <Box sx={{ flexGrow: 1, display: 'flex', height: '100%' }}>
          <Tabs
            orientation="vertical"
            variant="scrollable"
            value={selectedTab}
            onChange={handleSelectTab}
            aria-label="Vertical tabs"
            sx={{
              minWidth: 200,
              py: 4,
              borderRight: 1,
              borderColor: 'divider',
              '& .MuiTab-root': {
                alignItems: 'flex-end',
                pr: 4,
              },
            }}
          >
            <Tab label="Layout" />
            <Tab label="Color Scheme" />
            <Tab label="Font" />
          </Tabs>
          {selectedTab === 0 && (
            <TabPanel value={selectedTab} index={0}>
              <LayoutSettingsTab
                contentMode={contentMode}
                handleToggleContentMode={handleToggleContentMode}
              />
            </TabPanel>
          )}
          {selectedTab === 1 && (
            <TabPanel value={selectedTab} index={1}>
              <ColorSchemeSettingsTab
                darkMode={darkMode}
                syntaxHighlighterTheme={syntaxHighlighterTheme}
                handleToggleDarkMode={handleToggleDarkMode}
                setSyntaxHighlighterTheme={setSyntaxHighlighterTheme}
                
              />
            </TabPanel>
          )}
          {selectedTab === 2 && (
            <TabPanel value={selectedTab} index={2}>
              <FontSettingsTab
                fontSize={fontSize}
                fontFamily={fontFamily}
                fontFamilyMonospace={fontFamilyMonospace}
                fontInputMode={fontInputMode}
                fontInputModeMonospace={fontInputModeMonospace}
                setFontSize={setFontSize}
                setFontFamily={setFontFamily}
                setFontFamilyMonospace={setFontFamilyMonospace}
                handleFontInputModeChange={handleFontInputModeChange}
                handleFontInputModeMonospaceChange={handleFontInputModeMonospaceChange}
                
              />
            </TabPanel>
          )}
        </Box>
      </DialogContent>
      <DialogActions sx={{ px: 3, py: 2, borderTop: '1px solid', borderColor: 'divider' }}>
        <Button onClick={handleReset} size="small" sx={{ mr: 'auto', color: theme.palette.text.secondary }}>Restore default setting</Button>
        <Button onClick={handleCancel} size="small" sx={{ color: theme.palette.text.secondary }}>Cancel</Button>
        <Button onClick={handleSave} size="small">Save</Button>
      </DialogActions>
    </Dialog>
  );
};

export default SettingsDialog;
