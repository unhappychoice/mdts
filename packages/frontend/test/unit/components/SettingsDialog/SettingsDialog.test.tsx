import { fireEvent, render, screen, waitFor } from '@testing-library/react';
import { within } from '@testing-library/dom';
import React from 'react';
import { Provider } from 'react-redux';
import SettingsDialog from '../../../../src/components/SettingsDialog/SettingsDialog';
import { saveAppSetting } from '../../../../src/store/slices/appSettingSlice';
import { saveConfigToBackend } from '../../../../src/store/slices/configSlice';
import { createMockStore } from '../../../utils';

jest.mock('../../../../src/store/slices/appSettingSlice', () => ({
  ...jest.requireActual('../../../../src/store/slices/appSettingSlice'),
  saveAppSetting: jest.fn(() => ({ type: 'appSetting/saveAppSetting' })),
}));

jest.mock('../../../../src/store/slices/configSlice', () => ({
  ...jest.requireActual('../../../../src/store/slices/configSlice'),
  saveConfigToBackend: jest.fn(() => ({ type: 'config/saveConfigToBackend' })),
}));

describe('SettingsDialog', () => {
  let store;

  beforeEach(() => {
    store = createMockStore();
    store.dispatch = jest.fn();
  });

  const renderComponent = (open: boolean, onClose: () => void) => {
    return render(
      <Provider store={store}>
        <SettingsDialog open={open} onClose={onClose} />
      </Provider>,
    );
  };

  test('renders correctly when open', () => {
    const onClose = jest.fn();
    const { asFragment } = renderComponent(true, onClose);
    expect(asFragment()).toMatchSnapshot();
    expect(screen.getByText('Appearance Settings')).toBeInTheDocument();
  });

  test('does not render when closed', () => {
    const onClose = jest.fn();
    const { queryByText } = renderComponent(false, onClose);
    expect(queryByText('Appearance Settings')).not.toBeInTheDocument();
  });

  test('dispatches saveAppSetting and saveConfigToBackend on Save button click', async () => {
    const onClose = jest.fn();
    renderComponent(true, onClose);

    fireEvent.click(screen.getByText('Save'));

    await waitFor(() => {
      expect(saveAppSetting).toHaveBeenCalledWith({ darkMode: 'dark', contentMode: 'compact' });
      expect(saveConfigToBackend).toHaveBeenCalledWith({ fontFamily: 'Roboto', fontFamilyMonospace: 'monospace', fontSize: 14, syntaxHighlighterTheme: 'atom-dark' });
      expect(onClose).toHaveBeenCalledTimes(1);
    });
  });

  test('resets state and closes on Cancel button click', async () => {
    const onClose = jest.fn();
    renderComponent(true, onClose);

    // Change some settings
    fireEvent.click(screen.getByLabelText('full')); // Change content mode
    fireEvent.click(screen.getByRole('tab', { name: 'Color Scheme' }));
    await waitFor(() => {
      fireEvent.click(screen.getByLabelText('dark')); // Change dark mode
    });
    fireEvent.click(screen.getByRole('tab', { name: 'Font' })); // Select Font tab
    fireEvent.change(screen.getByRole('slider', { name: 'Font Size' }), { target: { value: 18 } }); // Change font size

    fireEvent.click(screen.getByText('Cancel'));

    expect(onClose).toHaveBeenCalledTimes(1);
    // Verify state is reset by checking if save was not called with changed values
    fireEvent.click(screen.getByText('Save')); // Try to save after cancel
    expect(saveAppSetting).toHaveBeenCalledWith({ darkMode: 'dark', contentMode: 'compact' });
    expect(saveConfigToBackend).toHaveBeenCalledWith({ fontFamily: 'Roboto', fontFamilyMonospace: 'monospace', fontSize: 14, syntaxHighlighterTheme: 'atom-dark' });
  });

  test('resets to default settings on Reset to Default button click', async () => {
    const onClose = jest.fn();
    renderComponent(true, onClose);

    // Change some settings
    fireEvent.click(screen.getByLabelText('full')); // Change content mode
    fireEvent.click(screen.getByRole('tab', { name: 'Color Scheme' }));
    await waitFor(() => {
      fireEvent.click(screen.getByLabelText('dark')); // Change dark mode
    });
    fireEvent.click(screen.getByRole('tab', { name: 'Font' })); // Select Font tab
    fireEvent.change(screen.getByRole('slider', { name: 'Font Size' }), { target: { value: 18 } }); // Change font size

    fireEvent.click(screen.getByText('Restore default setting'));

    await waitFor(() => {
      expect(saveAppSetting).toHaveBeenCalledWith({ darkMode: 'auto', contentMode: 'compact' });
      expect(saveConfigToBackend).toHaveBeenCalledWith({ fontFamily: 'Roboto', fontFamilyMonospace: 'monospace', fontSize: 14, syntaxHighlighterTheme: 'auto' });
    });
  });

  test('toggles dark mode', async () => {
    const onClose = jest.fn();
    renderComponent(true, onClose);

    fireEvent.click(screen.getByRole('tab', { name: 'Color Scheme' }));

    await waitFor(() => {
      fireEvent.click(screen.getByLabelText('dark'));
    });
    fireEvent.click(screen.getByText('Save'));
    expect(saveAppSetting).toHaveBeenCalledWith(expect.objectContaining({ darkMode: 'dark' }));

    fireEvent.click(screen.getByRole('tab', { name: 'Color Scheme' }));
    await waitFor(() => {
      fireEvent.click(screen.getByLabelText('light'));
    });
    fireEvent.click(screen.getByText('Save'));
    expect(saveAppSetting).toHaveBeenCalledWith(expect.objectContaining({ darkMode: 'light' }));

    fireEvent.click(screen.getByRole('tab', { name: 'Color Scheme' }));
    await waitFor(() => {
      fireEvent.click(screen.getByLabelText('auto'));
    });
    fireEvent.click(screen.getByText('Save'));
    expect(saveAppSetting).toHaveBeenCalledWith(expect.objectContaining({ darkMode: 'auto' }));
  });

  test('toggles content mode', () => {
    const onClose = jest.fn();
    renderComponent(true, onClose);

    fireEvent.click(screen.getByLabelText('full'));
    fireEvent.click(screen.getByText('Save'));
    expect(saveAppSetting).toHaveBeenCalledWith(expect.objectContaining({ contentMode: 'full' }));

    fireEvent.click(screen.getByLabelText('compact'));
    fireEvent.click(screen.getByText('Save'));
    expect(saveAppSetting).toHaveBeenCalledWith(expect.objectContaining({ contentMode: 'compact' }));
  });

  test('changes font size', async () => {
    const onClose = jest.fn();
    renderComponent(true, onClose);

    fireEvent.click(screen.getByRole('tab', { name: 'Font' }));

    await waitFor(() => {
      fireEvent.change(screen.getByRole('slider', { name: 'Font Size' }), { target: { value: 20 } });
    });
    fireEvent.click(screen.getByText('Save'));
    expect(saveConfigToBackend).toHaveBeenCalledWith(expect.objectContaining({ fontSize: 20 }));
  });

  test('changes font family via select', async () => {
    const onClose = jest.fn();
    renderComponent(true, onClose);

    fireEvent.click(screen.getByRole('tab', { name: 'Font' }));

    await waitFor(() => {
      fireEvent.click(within(screen.getByRole('radiogroup', { name: 'Font Family' })).getByText('Select from list'));
    });

    const select = screen.getByTestId('font-family-select');
    const button = within(select).getByRole('combobox');
    fireEvent.mouseDown(button);
    await waitFor(() => screen.getByRole('option', { name: 'Arial' }));
    fireEvent.click(screen.getByRole('option', { name: 'Arial' }));

    fireEvent.click(screen.getByText('Save'));
    expect(saveConfigToBackend).toHaveBeenCalledWith(expect.objectContaining({ fontFamily: 'Arial' }));
  });

  test('changes font family via text input', async () => {
    const onClose = jest.fn();
    renderComponent(true, onClose);

    fireEvent.click(screen.getByRole('tab', { name: 'Font' }));

    await waitFor(() => {
      fireEvent.click(within(screen.getByRole('radiogroup', { name: 'Font Family' })).getByText('Enter manually'));
    });
    const input = screen.getByTestId('font-family-input').querySelector('input') as HTMLInputElement;
    fireEvent.change(input, { target: { value: 'Custom Font' } });
    fireEvent.click(screen.getByText('Save'));
    expect(saveConfigToBackend).toHaveBeenCalledWith(expect.objectContaining({ fontFamily: 'Custom Font' }));
  });

  test('changes monospace font family via select', async () => {
    const onClose = jest.fn();
    renderComponent(true, onClose);

    fireEvent.click(screen.getByRole('tab', { name: 'Font' }));

    await waitFor(() => {
      fireEvent.click(within(screen.getByRole('radiogroup', { name: 'Monospace Font Family' })).getByText('Select from list'));
    });

    const select = screen.getByTestId('monospace-font-family-select');
    const button = within(select).getByRole('combobox');
    fireEvent.mouseDown(button);
    await waitFor(() => screen.getByRole('option', { name: 'Consolas' }));
    fireEvent.click(screen.getByRole('option', { name: 'Consolas' }));

    fireEvent.click(screen.getByText('Save'));
    expect(saveConfigToBackend).toHaveBeenCalledWith(expect.objectContaining({ fontFamilyMonospace: 'Consolas' }));
  });

  test('changes monospace font family via text input', async () => {
    const onClose = jest.fn();
    renderComponent(true, onClose);

    fireEvent.click(screen.getByRole('tab', { name: 'Font' }));

    await waitFor(() => {
      fireEvent.click(within(screen.getByRole('radiogroup', { name: 'Monospace Font Family' })).getByText('Enter manually'));
    });
    const input = screen.getByTestId('monospace-font-family-input').querySelector('input') as HTMLInputElement;
    fireEvent.change(input, { target: { value: 'Custom Monospace' } });
    fireEvent.click(screen.getByText('Save'));
    expect(saveConfigToBackend).toHaveBeenCalledWith(expect.objectContaining({ fontFamilyMonospace: 'Custom Monospace' }));
  });
});
