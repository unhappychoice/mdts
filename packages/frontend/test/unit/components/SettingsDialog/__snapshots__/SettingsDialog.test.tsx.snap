// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`SettingsDialog renders correctly when open 1`] = `<DocumentFragment />`;
